prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+not necessary that responder keep progress engine running.
+
+### 4.2 delivery complete
+
+The extra feature "delivery complete" was introduced with libfabric 1.12.0, and was assigned ID 1.
+
+Delivery complete is a requirement application can impose on an endpoint when opening the endpoint.
+It requires that when application gets the send/write completion, the application data must have
+been delivered to application's target buffer.
+
+The reason it is implemented as an extra feature is because, not all sub-protocols in the baseline
+features support delivery complete. Specifically, the following 6 sub-protocols do NOT:
+
+* eager message,
+* medium message,
+* long-CTS message,
+* eager write,
+* long-CTS write and
+* write atomic.
+
+These sub-protocols are designed to support a weaker completion model: transmit complete.
+Transmit complete requires that when the send/write completion was written, the data has been transmitted
+to the receiver/responder.
+
+The difference between transmit complete and delivery complete is transmit complete indicate
+that data has arrived at A buffer on the receive/responder, but the buffer is not necessary the application's
+target buffer. buffer. In fact, because of the limitation of the EFA device (no ordering guarantee) and the nature
+of the communications (emulated write), for some protocols, the implementation have to use a temporary
+buffer to receive data, and copy the data to application buffer later, and the time difference can be indefinite.
+
+The "delivery complete" extra feature was introduced to support applications with such requirements.
+It comes with 6 sub-protocols:
+
+* DC eager message,
+* DC medium message,
+* DC long-CTS message,
+* DC eager write,
+* DC long-CTS write and
+* DC write atomic.
+
+The workflow of these sub-protocols are same to that of there non-DC counterpart, with 3 differences changes:
+
+First, each DC capable sub-protocol defines its own REQ packet type.
+
+Second, after data was delivered to application buffer, the receiver/responder will send a RECEIPT
+packet back to the sender/requester.
+
+Third, sender/responder will not write completion until it received the RECEIPT packet.
+
+The binary format of a RECEIPT packet is as the following:
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the send operation |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `connid`         | 4 | integer | `uint32_t` | sender connection id, always present but optionally set. see section 4.4 for more detail |
+
+The `connid` field was introduced with the extra request "connid in header". Prior to that, it
```

Comment:
Per 3.1 -- this could potentially be something other than connid in the future, right? The key is knowing that in all cases there's at least a 4 byte padding already.

Reply:
Yes, you are right. In the future, this field can be used for other purpose, (with a different flag). I updated document for this.